File: 2025/Day 8/Aoc8.py
#!/usr/bin/env python3
import math
import sys
from itertools import combinations
from collections import Counter
from heapq import nsmallest
from typing import List, Tuple

class DSU:
    def __init__(self, n):
        self.p = list(range(n))
        self.r = [0]*n
        self.sz = [1]*n
    def find(self, a):
        p = self.p
        while p[a] != a:
            # path compression (iterative)
            p[a] = p[p[a]]
            a = p[a]
        return a
    def union(self, a, b):
        ra = self.find(a)
        rb = self.find(b)
        if ra == rb:
            return False
        if self.r[ra] < self.r[rb]:
            ra, rb = rb, ra
        self.p[rb] = ra
        self.sz[ra] += self.sz[rb]
        if self.r[ra] == self.r[rb]:
            self.r[ra] += 1
        return True
    def size(self, a):
        return self.sz[self.find(a)]

def read_points(path: str) -> List[Tuple[int,int,int]]:
    pts = []
    with open(path, "r") as f:
        for line in f:
            s = line.strip()
            if not s:
                continue
            x,y,z = map(int, s.split(","))
            pts.append((x,y,z))
    return pts

def squared_distance(a: Tuple[int,int,int], b: Tuple[int,int,int]) -> int:
    dx = a[0] - b[0]
    dy = a[1] - b[1]
    dz = a[2] - b[2]
    return dx*dx + dy*dy + dz*dz

def solve(path: str, k_pairs: int = 1000, use_heap_if_big: bool = True):
    pts = read_points(path)
    n = len(pts)
    if n < 2:
        print("Not enough points.")
        return 0

    total_pairs = n*(n-1)//2
    # If the total number of pairs is small-ish, build and sort all pairs.
    # Otherwise use heapq.nsmallest to avoid full sort if desired.
    # We'll use threshold: if total_pairs <= 5_000_000 (approx) it's fine to build full list in memory for typical inputs.
    threshold = 5_000_000

    if total_pairs <= threshold or not use_heap_if_big:
        pairs = []
        for i in range(n):
            xi, yi, zi = pts[i]
            for j in range(i+1, n):
                d = (xi-pts[j][0])**2 + (yi-pts[j][1])**2 + (zi-pts[j][2])**2
                pairs.append((d, i, j))
        # sort ascending by distance
        pairs.sort(key=lambda t: t[0])
        selected = pairs[:min(k_pairs, len(pairs))]
    else:
        # Use nsmallest to get k_pairs smallest pairs without storing all pairs
        # Generate pairs lazily
        def gen_pairs():
            for i in range(n):
                xi, yi, zi = pts[i]
                for j in range(i+1, n):
                    d = (xi-pts[j][0])**2 + (yi-pts[j][1])**2 + (zi-pts[j][2])**2
                    yield (d, i, j)
        selected = nsmallest(k_pairs, gen_pairs(), key=lambda t: t[0])

    # Now union the selected pairs in order
    dsu = DSU(n)
    # selected is already in ascending order; if we used nsmallest it might not be fully sorted,
    # so sort selected to ensure correct order processing.
    selected.sort(key=lambda t: t[0])
    for d, i, j in selected:
        dsu.union(i, j)

    # compute sizes
    roots = [dsu.find(i) for i in range(n)]
    counts = Counter(roots)
    sizes = sorted(counts.values(), reverse=True)

    # take three largest sizes; if less than 3 available, multiply what's there
    top3 = sizes[:3]
    # multiply
    prod = 1
    for s in top3:
        prod *= s

    print("Number of points:", n)
    print("Total pairs considered (selected):", len(selected))
    print("Top 3 component sizes:", top3)
    print("Answer (product of top 3):", prod)
    return prod

if __name__ == "__main__":
    # change filename if needed
    input_path = "input_day_8"
    solve(input_path)

#!/usr/bin/env python3
from typing import List, Tuple
import math

def read_points(path: str) -> List[Tuple[int,int,int]]:
    pts = []
    with open(path, "r") as f:
        for line in f:
            s = line.strip()
            if not s:
                continue
            x,y,z = map(int, s.split(","))
            pts.append((x,y,z))
    return pts

def sqdist(a: Tuple[int,int,int], b: Tuple[int,int,int]) -> int:
    dx = a[0] - b[0]
    dy = a[1] - b[1]
    dz = a[2] - b[2]
    return dx*dx + dy*dy + dz*dz

def prim_last_edge_xproduct(pts: List[Tuple[int,int,int]]):
    n = len(pts)
    if n < 2:
        raise ValueError("Need at least two points")

    in_tree = [False] * n
    best_dist = [10**30] * n   # large initial distance
    parent = [-1] * n

    # start from vertex 0
    best_dist[0] = 0
    parent[0] = -1

    # to store the maximum-weight MST edge (squared distance) and its endpoints
    max_edge_w = -1
    max_edge_u = -1
    max_edge_v = -1

    for _ in range(n):
        # pick next not-in-tree vertex with minimal best_dist
        u = -1
        u_dist = 10**30
        for i in range(n):
            if not in_tree[i] and best_dist[i] < u_dist:
                u_dist = best_dist[i]
                u = i
        if u == -1:
            break  # disconnected, shouldn't happen in complete graph

        # add u to the tree
        in_tree[u] = True

        # if u has a parent (i.e., not the initial root), consider its connecting edge
        if parent[u] != -1:
            w = best_dist[u]   # squared distance between parent[u] and u
            # track maximum MST edge weight and endpoints
            if w > max_edge_w:
                max_edge_w = w
                max_edge_u = parent[u]
                max_edge_v = u

        # update best_dist for remaining vertices using u as a new tree node
        for v in range(n):
            if in_tree[v]:
                continue
            d = sqdist(pts[u], pts[v])
            if d < best_dist[v]:
                best_dist[v] = d
                parent[v] = u

    # after MST built, max_edge_u and max_edge_v are the endpoints of the largest MST edge
    if max_edge_u == -1 or max_edge_v == -1:
        raise RuntimeError("No MST edge found (unexpected)")

    # multiply X coordinates
    x_prod = pts[max_edge_u][0] * pts[max_edge_v][0]
    return x_prod, (max_edge_u, max_edge_v, max_edge_w)

if __name__ == "__main__":
    input_path = "input_day_8"   # change if needed
    pts = read_points(input_path)
    prod, info = prim_last_edge_xproduct(pts)
    u, v, w = info[0], info[1], info[2]
    print("Index endpoints:", info[0], info[1])
    print("Squared distance of that edge:", info[2])
    print("Product of X coordinates:", prod)

File: 2025/Day 8/Aoc8.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strconv"
	"strings"
)

// Point represents a 3D point
type Point struct {
	x, y, z int
}

// DSU (Disjoint Set Union / Union-Find)
type DSU struct {
	p  []int
	r  []int
	sz []int
}

func NewDSU(n int) *DSU {
	dsu := &DSU{
		p:  make([]int, n),
		r:  make([]int, n),
		sz: make([]int, n),
	}
	for i := 0; i < n; i++ {
		dsu.p[i] = i
		dsu.sz[i] = 1
	}
	return dsu
}

func (dsu *DSU) find(a int) int {
	// Path compression
	for dsu.p[a] != a {
		dsu.p[a] = dsu.p[dsu.p[a]]
		a = dsu.p[a]
	}
	return a
}

func (dsu *DSU) union(a, b int) bool {
	ra := dsu.find(a)
	rb := dsu.find(b)
	if ra == rb {
		return false
	}
	if dsu.r[ra] < dsu.r[rb] {
		ra, rb = rb, ra
	}
	dsu.p[rb] = ra
	dsu.sz[ra] += dsu.sz[rb]
	if dsu.r[ra] == dsu.r[rb] {
		dsu.r[ra]++
	}
	return true
}

func (dsu *DSU) size(a int) int {
	return dsu.sz[dsu.find(a)]
}

// PairDist represents a pair with its squared distance
type PairDist struct {
	dist int
	i, j int
}

func readPoints(path string) ([]Point, error) {
	file, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	var points []Point
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" {
			continue
		}
		parts := strings.Split(line, ",")
		if len(parts) != 3 {
			continue
		}
		x, _ := strconv.Atoi(strings.TrimSpace(parts[0]))
		y, _ := strconv.Atoi(strings.TrimSpace(parts[1]))
		z, _ := strconv.Atoi(strings.TrimSpace(parts[2]))
		points = append(points, Point{x, y, z})
	}
	return points, scanner.Err()
}

func squaredDistance(a, b Point) int {
	dx := a.x - b.x
	dy := a.y - b.y
	dz := a.z - b.z
	return dx*dx + dy*dy + dz*dz
}

func solvePart1(path string, kPairs int) int {
	pts, err := readPoints(path)
	if err != nil {
		fmt.Printf("Error reading file: %v\n", err)
		return 0
	}

	n := len(pts)
	if n < 2 {
		fmt.Println("Not enough points.")
		return 0
	}

	var selected []PairDist

	// Build all pairs and sort
	totalPairs := n * (n - 1) / 2
	pairs := make([]PairDist, 0, totalPairs)
	for i := 0; i < n; i++ {
		for j := i + 1; j < n; j++ {
			d := squaredDistance(pts[i], pts[j])
			pairs = append(pairs, PairDist{d, i, j})
		}
	}
	sort.Slice(pairs, func(i, j int) bool {
		return pairs[i].dist < pairs[j].dist
	})
	if len(pairs) < kPairs {
		selected = pairs
	} else {
		selected = pairs[:kPairs]
	}

	// Union the selected pairs
	dsu := NewDSU(n)
	for _, pair := range selected {
		dsu.union(pair.i, pair.j)
	}

	// Count component sizes
	roots := make(map[int]int)
	for i := 0; i < n; i++ {
		root := dsu.find(i)
		roots[root]++
	}

	// Get top 3 sizes
	sizes := make([]int, 0, len(roots))
	for _, count := range roots {
		sizes = append(sizes, count)
	}
	sort.Slice(sizes, func(i, j int) bool {
		return sizes[i] > sizes[j]
	})

	top3 := sizes
	if len(sizes) > 3 {
		top3 = sizes[:3]
	}

	// Calculate product
	prod := 1
	for _, s := range top3 {
		prod *= s
	}

	fmt.Printf("Number of points: %d\n", n)
	fmt.Printf("Total pairs considered (selected): %d\n", len(selected))
	fmt.Printf("Top 3 component sizes: %v\n", top3)
	fmt.Printf("Answer (product of top 3): %d\n", prod)

	return prod
}

func solvePart2(path string) int {
	pts, err := readPoints(path)
	if err != nil {
		fmt.Printf("Error reading file: %v\n", err)
		return 0
	}

	n := len(pts)
	if n < 2 {
		fmt.Println("Need at least two points")
		return 0
	}

	// Prim's algorithm to find MST and track largest edge
	inTree := make([]bool, n)
	bestDist := make([]int, n)
	parent := make([]int, n)

	// Initialize
	const largeNum = 1000000000
	for i := 0; i < n; i++ {
		bestDist[i] = largeNum
		parent[i] = -1
	}

	// Start from vertex 0
	bestDist[0] = 0

	maxEdgeW := -1
	maxEdgeU := -1
	maxEdgeV := -1

	// Build MST
	for iter := 0; iter < n; iter++ {
		// Find next vertex with minimum distance not yet in tree
		u := -1
		uDist := largeNum
		for i := 0; i < n; i++ {
			if !inTree[i] && bestDist[i] < uDist {
				uDist = bestDist[i]
				u = i
			}
		}

		if u == -1 {
			break
		}

		// Add u to tree
		inTree[u] = true

		// Track maximum edge
		if parent[u] != -1 {
			w := bestDist[u]
			if w > maxEdgeW {
				maxEdgeW = w
				maxEdgeU = parent[u]
				maxEdgeV = u
			}
		}

		// Update distances for remaining vertices
		for v := 0; v < n; v++ {
			if inTree[v] {
				continue
			}
			d := squaredDistance(pts[u], pts[v])
			if d < bestDist[v] {
				bestDist[v] = d
				parent[v] = u
			}
		}
	}

	if maxEdgeU == -1 || maxEdgeV == -1 {
		fmt.Println("No MST edge found")
		return 0
	}

	// Calculate product of X coordinates
	xProd := pts[maxEdgeU].x * pts[maxEdgeV].x

	fmt.Printf("Index endpoints: %d %d\n", maxEdgeU, maxEdgeV)
	fmt.Printf("Squared distance of that edge: %d\n", maxEdgeW)
	fmt.Printf("Product of X coordinates: %d\n", xProd)

	return xProd
}

func main() {
	// Get current working directory
	cwd, err := os.Getwd()
	if err != nil {
		fmt.Printf("Error getting current directory: %v\n", err)
		return
	}

	// Construct path to input file
	inputPath := filepath.Join(cwd, "input_day_8")

	fmt.Println("=== Day 8 - Part 1 ===")
	solvePart1(inputPath, 1000)

	fmt.Println("\n=== Day 8 - Part 2 ===")
	solvePart2(inputPath)
}
